Refactor credentials and new package `core`

VictorVelarde opened the pull request:
Related to #17 

This PR introduces @carto/toolkit-core, as a way to host some common pieces of code, to be shared and present in most projects. 

The first piece is Credentials, removing the previous duplication, in which every package managed those locally.

jesusbotella commented on packages/custom-storage/src/sql/SQLStorage.ts:
```diff
@@ -79,10 +79,8 @@ export class SQLStorage {
-    const client = sqlClient || this._sql;
-
-    return client.query(`
+  public getVisualizations(): Promise<StoredVisualization[]> {
```
I do agree that it's better now, but I think that the parameter was meant for using a different SQL client with different credentials, and thus different permissions levels.

VictorVelarde replied:
I also thought about testing... who knows.

Managing a different sqlClient injected by param felt strange to me there. If we find the need we can re-add it in the future, but it would probably make sense just to create another SQLStorage instance

jesusbotella commented on packages/maps/__tests__/Client.spec.ts:
```diff
@@ -0,0 +1,30 @@
+    expect(() => {
+      // tslint:disable-next-line: no-unused-expression
+      new Client('', '');
+    }).toThrow('Username is required');
```
What about returning the expression? 😁
I guess that this would have the same result and allows us to remove the tslint override

```js
expect(() => new Client('', '')).toThrow('Username is required');
```

VictorVelarde replied:
Clearly better. Nice tip! 👍

jesusbotella commented on packages/maps/__tests__/Client.spec.ts:
```diff
@@ -0,0 +1,30 @@
+    expect(m).toBeTruthy();
+  });
+
+  it('requires credentials', () => {
```
Just in case we want to be safer, we should check that it throws when API Key is missing as well.

VictorVelarde replied:
Added a little new test. If we add a Credentials class to ctor, then we will avoid also these duplicities

VictorVelarde commented on packages/core/src/Credentials.ts:
```diff
@@ -0,0 +1,55 @@
+  }
+
+  public static get DEFAULT_PUBLIC_API_KEY() {
+    return  DEFAULT_PUBLIC_API_KEY;
```
Thx, those new 👓 are really working 😄 !

VictorVelarde replied:
Aggree on the idea about 'better unified creds'. We should improve the way we managed them between packages. 

Do you think that we should start by just passing a Credentials object to main constructors (I thought about it but I finally didn't do it):
- maps --> Maps class at Client.ts
- sql --> SQL class at Client.ts
- custom-storage --> CustomStorage at CustomStorage.ts. By the way, we could probably rename it also to Client.ts for consistency (and maybe even reconsider about exporting PublicSQLReader)

VictorVelarde commented on packages/core/src/index.ts:
```diff
@@ -0,0 +1,3 @@
+import Credentials from './Credentials';
+
+export { Credentials };
```
Yeah, in the previous state I had 2 imports. Consider it done!

VictorVelarde commented:
I have applied all your suggestions, thx! 👍 

I prefer https://github.com/CartoDB/toolkit/pull/58. to get merged first, and I will adapt this to master after that. 

In that moment maybe I can adapt the ctors too, as commented on https://github.com/CartoDB/toolkit/pull/57#discussion_r375419754

VictorVelarde commented:
I have rebased and adapted to master this PR. I added also Credentials as param to main constructors, instead of separated components (username, apiKey...)

VictorVelarde commented:
Are you ok with merging this in @neokore @jesusbotella ?

jesusbotella commented:
Yes!

neokore commented:
Sure!
